### ISSUE TITLE
IList<T> optimizations become deoptimizations for arrays and array wrappers

### ISSUE THREAD

**Issue by stephentoub** [dotnet/runtime#101360]
https://github.com/dotnet/runtime/issues/62457 highlights how `T[]`s interface implementations can't be devirtualized, because the interface implementations come from the runtime. However, `T[]`'s enumerator implementation is implemented in managed code, and thus can be devirtualized. This creates an inversion of performance for arrays or for types that wrap arrays but return the array's enumerator as its enumerator, e.g. a `new ReadOnlyCollection<T>(someArray)`. We have a bunch of places that do things like:
```C#
if (collection is IList<T> list)
{
    int count = list.Count;
    for (int i = 0; i < count; i++) Process(list[i]);
}
else
{
    foreach (var item in collection) Process(item);
}
```
which avoids an enumerator allocation and in theory should drop two interface dispatches per element down to one. However, for a case such as `collection = new ReadOnlyCollection<T>(someArray)`, the ROC's `GetEnumerator` just returns the arrays' enumerator, which can be devirtualized with PGO. That means that while the `IList<T>` path does avoid the enumerator allocation, it can end up being measurably slower than the fallback enumerable path, because the enumerable path may be able to devirtualize and inline both interface calls to the enumerator's MoveNext / Current, whereas it's unable to for the array behind the ROC.

| Method        | Mean     | Allocated |
|-------------- |---------:|----------:|
| SumList       | 2.379 us |         - |
| SumEnumerable | 1.342 us |      32 B |

```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System.Collections.ObjectModel;

BenchmarkSwitcher.FromAssembly(typeof(Tests).Assembly).Run(args);

[MemoryDiagnoser]
[HideColumns("Job", "Error", "StdDev", "Median", "RatioSD")]
public class Tests
{
    private ReadOnlyCollection<int> _source = new ReadOnlyCollection<int>(Enumerable.Range(0, 1000).ToArray());

    [Benchmark]
    public int SumList()
    {
        ReadOnlyCollection<int> source = _source;
        int sum = 0;
        int count = source.Count;
        for (int i = 0; i < count; i++)
        {
            sum += source[i];
        }
        return sum;
    }

    [Benchmark]
    public int SumEnumerable()
    {
        int sum = 0;
        foreach (int item in _source)
        {
            sum += item;
        }
        return sum;
    }
}
```

cc: @egorbo

**Comment by jkotas** [comment]
The manual `IList<T>` special-casing optimization is based on an assumption that iterating all elements using an `IList<T>` indexer has better performance than `IList<T>` enumerator. There are many different reasons why this assumption can be invalid for a specific `IList<T>` implementation. This problem exists for ReadOnlyCollection even without PGO and de-virtualization in the picture.

Do you have a solution in mind?

**Comment by jkotas** [comment]
> Can you elaborate?

The de-optimization will occur for any IList<T> implementation that has relatively more expensive indexer than enumerator.  For example, the manual IList<T> special-casing is going to be a de-optimization for `ImmutableList<T>` too. It is likely that any `IList<T>` implementation that is not directly backed by an in-memory array is going to have relatively more expensive indexer than enumerator.

> Fixing https://github.com/dotnet/runtime/issues/62457 would seem to address it.

Yes, it will help in the specific case of ReadOnlyCollection wrapper over arrays where the callsites are not polymorphic and devirtualization is able to kick in. It helps because it shifts the relative costs of the indexer and enumerator.

I believe that #62457 is fixable, it requires cleaning up the hacks used to implement the generic array interfaces.

**Comment by MichalPetryka** [comment]
> I believe that #62457 is fixable, it requires cleaning up the hacks used to implement the generic array interfaces.

I recall @MichalStrehovsky saying that it's not an issue in the NativeAOT VM, maybe it'd be worth to benchmark things there too.

**Comment by jkotas** [comment]
Yes, NativeAOT is able devirtualize the generic collection interface methods on arrays without a problem. (Porting the native AOT implementation strategy would be one way to fix https://github.com/dotnet/runtime/issues/62457.)

**Comment by stephentoub** [comment]
> The de-optimization will occur for any IList implementation that has relatively more expensive indexer than enumerator. 

Sure, and we've had that discussion in the past specifically about `ImmutableList<T>`; it's one of the few (if not the only) `IList<T>`s we ship whose indexer is not O(1). In general we've avoided making such types `IList<T>`, e.g. `LinkedList<T>` is not `IList<T>`, and ideally it wouldn't have either. But that ship sailed.

I was more asking about the "This problem exists for ReadOnlyCollection even without PGO and de-virtualization in the picture" statement, as with ROC this very much did seem to be about PGO; without PGO, indexer <= enumerator, but with PGO, indexer may be > enumerator.

**Comment by jkotas** [comment]
> as with ROC this very much did seem to be about PGO; without PGO, indexer <= enumerator, but with PGO, indexer may be > enumerator.

I agree, with the emphasis on "may be". If the callsites are monomorphic, indexer wins. If the callsites are polymorphic, enumerator can still be a winner.

**Comment by stephentoub** [comment]
Closing as addressed in 10 by @AndyAyersMS
